Iss766, part 3: adding MetaStore class

brendan-m-murphy opened the pull request:
* **Summary of changes** (Bug fix, feature, docs update, ...)

The main change is a new `MetaStore` interface (ABC) as well as an implementation for TinyDB.
All dependency on `TinyDB` is now in the `objectstore.metastore` module. This means that changing
the metastore backend only requires code changes in `objectstore.metastore`, provided the backend
can implement the `MetaStore` interface, which has four abstract methods (basically "CRUD").

The `TinyDB` implementation used throughout the code base is `ClassicMetaStore`, which has the same
structure as the current (v6.2) metastore. This inherits from `TinyDBMetaStore`, since the caching middleware
requires a context manager, which isn't a general requirement for `TinyDB`. The `ClassicMetaStore` can be
accessed via `open_metastore`, which is a context manager, or via the class method `ClassicMetaStore.from_bucket`,
which requires that the metastore is closed at a later point. The `from_bucket` class method is necessary since
`BaseStore` operates via a context manager.

I'm not sure of the best way to initialise `MetaStore` objects; it will depend on how they're called and what other
backends require. Context managers would allow us to support "transactions" in Mongodb. They're also necessary
for caching middleware in TinyDB, since it won't write to disk until closed, unless more than 1000 writes have occurred.

All `MetaStore` classes implement:
- `search` (in TinyDB, using `search(Query().fragment(...))`)
- `insert` (in TinyDB, using `insert`)
- `update` (in TinyDB, using `update` for adding/overwriting as well as deleting keys)
- `delete` (in TinyDB, using `remove`)

Also, there is a `select` method implementing using `search`, which can be used to retrieve all values for a given key.
For instance: `metastore.select("uuid")` retrieves all UUIDs.

The `TinyDBMetaStore` also formats all metadata/search terms so that the keys are lower case.

Other changes include:
- removed dependency on `BaseStore._datasource_uuids` and `BaseStore.datasources()`. This partially duplicates
   changes in PR #752, and leaves the `objectstore._integrity` module in a less-than-ideal state. Since `MetaStore` is 
   oblivious to `Datasource`, I don't think we can support tracking datasource uuids outside the metastore, unless we want
   to call `BaseStore` (children) to do this. (Update: PR #752 has been accepted and merged into this PR.)
- deleted `store._metadata` module; some code moved to `ClassicMetaStore`, and the `data_manager` function moved to the `dataobjects._datamanager` module.
- unit tests for the metastore added. These could be parameterised if we add other backends.
- context managers removed from `BaseStore` children; they just inherit the `BaseStore` version. 
- changes required to remove dependencies from `TinyDB` outside of `objectstore.metastore`: deleted dead
   code from `retrieve._search`; changes `datasource_lookup` and `assign_data` in `BaseStore`; changes to `Datamanager` and `integrity_check`; changes to tests.

No changes were made to the API (`standardise`, `search`, `analyse`, etc.)

* **Please check if the PR fulfills these requirements**

- [x] Closes #766 
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [x] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [x] Documentation and tutorials updated/added
- [x] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature
- [x] Added any new requirements to `requirements.txt` and `recipes/meta.yaml`

rt17603 commented on openghg/objectstore/metastore/_metastore.py:
```diff
@@ -0,0 +1,242 @@
+        pass
+
+    @abstractmethod
+    def add(self, metadata: MetaData) -> None:
```
In general, the meaning of *add* is a bit loose in database design (not fully agreed). For the TinyDB implementation below the insert function has been used which does have a specific meaning - "add this, but error if it already exists" (which contrasts with other *add* type options "update" which means "replace the existing thing, and error it if *doesn't* exist" and "upsert" which is a combination which means "update this thing if it exists, or add it if it doesn't") - [TinyDB's definitions](https://tinydb.readthedocs.io/en/latest/usage.html) and this is also consistent with SQL.

This is a long-winded way of saying, I think the name `insert` is better here, but would also require an explicit comment in the docstring saying "it will error it it already exists". That way, when adding different subclasses in the future, the interface is explicitly defined.

brendan-m-murphy replied:
Sure sounds good. I can't require the exception in the interface but I will mention it in the docstring.

brendan-m-murphy replied:
I don't think TinyDB's insert will throw an error unless the Document you're adding already has a document id, and this document id is already in the database. The analogous behaviour for the object store is to throw an error if we try to add/insert metadata that already has a UUID, and this UUID is already used.

So, I'll make the name change, but I won't add the insert/update error to the metastore at this point.

(Somewhat related to this: I moved some error handling to the ABC, since it only depends on `search`.)

One option is to make a mixin if we want strict inserts:
`
class StrictInsertMixin:
    def insert(self, metadata):
        if len(super().search(metadata)) > 1:
            raise MetastoreError
        super().insert(metadata)
`

brendan-m-murphy replied:
I changed the name of the method to `insert`. I think if we want to raise errors on inserting metadata that matches existing metadata (ignoring document IDs), the mixin is a decent solution.

At some point, we will need to create `MetaStore`s based on configs set in a bucket, so no one will manually have to type the name `StrictTinyDBMetaStore`, they can just call `open_metastore`.

rt17603 replied:
Ok so with the current design using insert through TinyDB multiple times with the same UUID would not throw an exception because UUID is not identified as the Document ID? And it looks like we can't set the Document ID to be the UUID because this needs [to be an integer](https://tinydb.readthedocs.io/en/latest/api.html#tinydb.table.Document).

I feel like the Mixin solution may be overly complex - it would be fine to make it clear what we expect the child class to do within the documentation. We could add this check to the TinyDB insert method to make sure this is enforced.

brendan-m-murphy replied:
Yes with the current design, you can insert the same data as many times as you like (each instance would have a different document ID, created by TinyDB) unless you specify the document ID as one that already exists.

The metastore isn't responsible for UUIDs at all, that's done currently by `BaseStore` and will be done by `ObjectStore` in the future. I think it makes more sense for UUIDs to be handled at the "object store" level, since it is information needed by both the metastore and datasources.

The metastore ABC is just meant to be an abstraction over the database, and the simpler the better, provided we can do everything we need with the operations it provides.

I can add a check to require that the particular set of key-value pairs is unique when data is inserted. To be consistent, we would have to add checks to make sure that updating metadata doesn't cause it to duplicate any existing data.

I added the checks to update and delete because I think accidentally deleting a lot of data would be annoying, and I put a requirement that only one item is updated at a time, since the logic for updating multiple items is significantly more complex, and we could just use a for loop to do it if we need to. 

At the object store level, we can manage which keys are immutable, so it won't be possible to update the uuid of an item, and so updating can never duplicate another record. 

TLDR: ensuring metastore entries are unique will be easier to do within the object store.

I'll make a draft PR with the `ObjectStore` class so you can see what I mean.

brendan-m-murphy replied:
This is the branch with the `ObjectStore` class: iss766_part_4_objectstore_with_ABC_datasource

rt17603 reviewed (changes requested):
This looks really good - a really nice and clear implementation which clearly signposts how this can be built upon in future. Also doing a great job in starting to compartmentalise the layers of functionality.

Just a few very minor comments / requested changes to address before this is ready.

rt17603 commented on openghg/objectstore/metastore/_classic_metastore.py:
```diff
@@ -0,0 +1,159 @@
+    return f"{result['_root']}/uuid/{result['_uuid']}/metastore"
+
+
+class ObjectStorage(tinydb.Storage):
```
I was more thinking that this object is tied specifically to the Metastore at the moment but "ObjectStorage" as a term to me describes the wider concept of the store encompassing the Metastore and the Data itself?

gareth-j replied:
Yeah but it is for storing the TinyDB in the object store.

rt17603 replied:
Yes I agree on your TinyDB point - I'm more querying the "ObjectStore" name - what does that mean to us when we're defining our system?

brendan-m-murphy replied:
Maybe `BucketKeyStorage` since its subclassing `Storage` and the main difference from just writing to a particular file is that it's determining where to store things given a bucket and a key.

brendan-m-murphy replied:
Also, I just kept the old name, so I didn't really choose `ObjectStorage` with some particular intent. I'm happy to change it.

brendan-m-murphy replied:
I went with `BucketKeyStorage`. I think because it subclasses `tinydb.Storage` and it's in a module that implements a subclass of `TinyDBMetaStore`, I don't we need to include "TinyDB" in the name.

rt17603 replied:
Sounds good - can come back to if we need later on.

brendan-m-murphy commented on openghg/objectstore/metastore/_metastore.py:
```diff
@@ -0,0 +1,242 @@
+
+from openghg.types import MetastoreError
+
+MetaData = Dict[str, Any]
```
Yea I guess it's not helpful for the user, but it makes the code a bit clearer (in terms of intent). The type aliases could be replaced by actual classes at some point (this would be good for MetaData, not sure about the others).

brendan-m-murphy replied:
I've left these as is. Besides "MetaData" and "Data", I think most of these will remain internal to the objectstore module, and they could be refined in the future.

For the public API (standardise, etc.) I think the type aliases are a bad idea.

brendan-m-murphy commented on tests/store/test_obssurface.py:
```diff
@@ -834,7 +837,7 @@ def test_store_icos_carbonportal_data(bucket, mocker):
     ],
 )
-def test_obs_schema(species, obs_variable):
+def test_metastore_schema(species, metastore_variable):
```
Sorry, this was probably an errant find-and-replace, I'll put it back.

brendan-m-murphy replied:
I put this back to how it was (and switched to using line-by-line search and replace....)

brendan-m-murphy commented on tests/store/test_obssurface.py:
```diff
@@ -537,10 +539,12 @@ def test_read_thames_barrier(bucket):
+        assert sorted(uuids) == expected_keys
 
 
+@pytest.mark.xfail(reason="Deleting datasources will be handled by ObjectStore objects")
```
`ObjectStore` is planned for the next version. This test doesn't really belong here, but currently `ObsSurface` has a delete method and no other classes in the `BaseStore` family do.

brendan-m-murphy replied:
This is still xfail, but I added a TODO comment to move this test in Part 4 (related to issue #727)

rt17603 replied:
In which case - linking to the issue would be great
```suggestion
@pytest.mark.xfail(reason="Deleting datasources will be handled by ObjectStore objects - links to issue #727")
```

brendan-m-murphy replied:
I added a link to issue #727

brendan-m-murphy commented on tests/objectstore/test_classic_metastore.py:
```diff
@@ -0,0 +1,32 @@
+from openghg.types import MetastoreError
+
+
+def test_metastore_read_write_mode(tmp_path):
```
I can put the path in with the others, but the particular path isn't really important for these tests.

We might want to use the pytest fixtures rather than tempfile, since the pytest ones persist after the tests (until the next test run?) whereas the tempfile deletes them immediately.

rt17603 replied:
I just think for the tests we don't want to be creating temporary files/locations all over the user's system - I don't mind where this is done. If creating this using in build `pytest` features like `tmp_path` then that's fine but we don't want to approach this in 2 different ways really.

brendan-m-murphy replied:
I haven't addressed this in the latest PR. For unit tests on isolated components, I think `tmp_path` makes sense.

For more involved tests, mocking the actual OpenGHG set up makes sense. Keeping the file path names in the top level conftest,py makes sense as well, although creating the temporary directories at the module level would mean that we can run the tests in parallel. (Any tests that don't actually modify the object store could use data added at the session level though.)

brendan-m-murphy replied:
> I just think for the tests we don't want to be creating temporary files/locations all over the user's system - I don't mind where this is done. If creating this using in built `pytest` features like `tmp_path` seems like the best approach then that's fine but we don't want to implement this in 2 different ways really.

Sorry I didn't see this before I commented. You can set the location where pytest puts temp paths and directories; I don't know if I can make it the same as where the test stores are kept.

If I switch over the existing conftest files to use `tmp_path_factory` (which makes directories that persist with session level scope), then we'd have the option to use `tmp_path` and still keep all of the temporary files in the same location.

The files created are labelled pytest-NUM where NUM is incremented with each run. These get cleaned up eventually, and only a few persist at a time. This sounds useful to me because sometimes to debug things I have to put in a breakpoint, then find the temporary path, then look at the temporary directory while the debugger is running.

It is possible to only keep at most one temporary directory at a time: https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#base-temporary-directory

To avoid using `tmp_path` for this PR, I could create a function scope fixture in the top level conftest.py that provides a directory in the `tempfile` temp dir, and then clears it after the test runs. That would do the same thing that `tmp_path` is doing.

I think refactoring the tests is outside the scope of this PR, but another option is to leave in `tmp_dir` and open an issue to refactor the tests to use `tmp_dir` and `tmp_dir_factory` instead of pathlib and tempfile. (I'm not super keen on doing this, but it would be useful for unit tests where we want to test things in isolation. The other option for doing this is making an in-memory implementation of the interface for testing. In this case, `tmp_dir` was easier.)

brendan-m-murphy commented on tests/store/test_obssurface.py:
```diff
@@ -31,13 +32,12 @@ def test_raising_error_doesnt_save_to_store(mocker, bucket):
-            # with obs that results in an exception being raised that isn't internal
-            # to our processing functions
-            raise ValueError("Oops")
+        key = "abc123"
```
Okay I'll rework this.

brendan-m-murphy replied:
I put back the `open_metastore` context, but kept the dummy key "abc123". Before, `ObsSurface.key()` was used, but the point is that anything besides `key = ObsSurface.metakey()` should fail (I think).

brendan-m-murphy commented on openghg/objectstore/_integrity.py:
```diff
@@ -1,33 +1,36 @@
                     Datasource.load(bucket=bucket, uuid=uid, shallow=True).integrity_check()
 
-                metastore_uuids = [r["uuid"] for r in sc._metastore]
+                metastore_uuids = metastore.select("uuid")
```
This is the part that needs to be changed to match PR #752. I'll add Gareth's new version.

brendan-m-murphy commented on openghg/objectstore/metastore/_classic_metastore.py:
```diff
@@ -0,0 +1,159 @@
+from openghg.types import MetastoreError
+
+
+object_store_data_classes = {
```
I think openghg.store.spec needs to be changed because it creates circular dependencies (e.g. with objectstore._integrity), which are hidden by importing within functions instead of at top level.

Ideally there would be a config subpackage that configures object stores, and this info should probably be read from a config file. After the next PR in the sequence, it might be the case that only things within objectstore need to know about these keys. 

One solution is to put store.spec outside of store for now, or to wait until the next PR in the sequence to see if these don't need to be outside objectstore.

A better solution would be to have a different tables for each data type inside a single metastore. This is supported by TinyDB and MongoDB, but I didn't want to do it now because it's too much of a change from the current set up.

brendan-m-murphy replied:
I didn't resolve this, but I added a TODO comment to move this code in the next iteration (Part 4, related to issue #727 ). I suspect that after `BaseStore` is refactored, the only place this info will be needed is in the `objectstore` module. (If I do it correctly, this should be the case, since it is info internal to the object store.)

rt17603 replied:
> I think openghg.store.spec needs to be changed because it creates circular dependencies (e.g. with objectstore._integrity), which are hidden by importing within functions instead of at top level.

Ok - one comment on the circular imports is that I did find with careful ordering of imports and only importing exactly where needed I was able to resolve these when they came up in code.

> One solution is to put store.spec outside of store for now, or to wait until the next PR in the sequence to see if these don't need to be outside objectstore.

Yes it would be fine for the details to be in a separate sub-package if needed

> A better solution would be to have a different tables for each data type inside a single metastore. This is supported by TinyDB and MongoDB, but I didn't want to do it now because it's too much of a change from the current set up.

Yes, I think we would need to discuss if that was the best solution so we don't want to currently make this change (at least at the moment)

> I didn't resolve this, but I added a TODO comment to move this code in the next iteration (Part 4, related to issue https://github.com/openghg/openghg/issues/727 ). I suspect that after BaseStore is refactored, the only place this info will be needed is in the objectstore module. (If I do it correctly, this should be the case, since it is info internal to the object store.)

Ok that's fine as long as this is tracked. Would be good to add this as a comment on Issue #766 so it's clear this still needs to be done.

brendan-m-murphy replied:
I referenced this discussion on Issue #727. (I think Issue #766 can be closed with this PR, since 766 is just splitting off the metastore from `BaseStore`. I guess "part 4" is a misnomer, since the next PR in the sequence will address #727.)

brendan-m-murphy commented on tests/objectstore/test_integrity.py:
```diff
@@ -69,15 +57,15 @@ def test_integrity_check_delete_Datasource_keys():
         integrity_check()
 
-
+@pytest.mark.xfail(reason="metastore.datasources() now returns metastore uuids")
```
I'm going to change this to match what Gareth has done in PR #752

brendan-m-murphy replied:
This is no longer xfail, it's been changed to what Gareth did in PR #752

brendan-m-murphy commented on tests/store/test_obssurface.py:
```diff
@@ -234,7 +234,7 @@ def test_read_CRDS(bucket):
     assert data_keys == new_expected_keys
 
-
+@pytest.mark.xfail(reason="The values stored with _datasource_uuids are changing.")
```
This will follow the changes in PR #752

brendan-m-murphy replied:
This is no longer xfail, it's been changed to what Gareth did in PR #752

brendan-m-murphy commented on openghg/objectstore/metastore/_metastore.py:
```diff
@@ -0,0 +1,242 @@
+
+        Args:
+            search_terms: dictionary of key-value pairs to search by.
+
```
I guess the google style is to have empty lines before Args, Returns, and Raises: https://google.github.io/styleguide/pyguide.html#doc-function-args (also, no type hints in the docstring if type annotations are used, and no return if first line is sufficient, or return type is None)

We don't need to have empty lines though (or do the other google docstring stuff). I guess I'm doing the "raises" part wrong according to the link.

Maybe in the developer docs we can write out what we want, either google style verbatim, or google style with X changes (or just write out examples and just say "similar" to google style).

(Also, I haven't got my text editor to auto insert docstrings yet... I used to have it. Does VSCode automatically do the style you prefer?)

rt17603 replied:
I think in general it doesn't matter which style we adopt as long as we're clear and we do this consistently.

We should make sure that it's clear what style we want to use and that everyone is on board. In the current documentation I can see we have this (put together by Gareth): https://github.com/openghg/openghg/blob/devel/doc/source/development/python_devel.rst#coding-style (https://github.com/openghg/openghg/blob/devel/doc/source/contributing.rst#contributing-new-code has a broken link to presumably this page as well?). This includes some examples of doc-strings but not necessarily in the style being applied here.

My suggestion would be that, if we want to apply a style as has been requested here this should be in this style guide. That way this can be referenced (and agreed) when having these discussions.

rt17603 replied:
To that end I've opened up an issue - https://github.com/openghg/openghg/issues/782. This may seem trivial but I think it's important we get this agreed so we can help us easily keep a consistent style (less review in PRs as well) - @gareth-j @brendan-m-murphy Can one of you take this on and then we can finalise this? Hopefully shouldn't take too long to just get this written down.

brendan-m-murphy replied:
Sorry I forgot that the was documentation about this already

brendan-m-murphy commented on openghg/objectstore/metastore/_classic_metastore.py:
```diff
@@ -0,0 +1,159 @@
+        self._mode = mode
+
+    def read(self) -> Optional[dict]:
+        """Read data from DB.
```
Updated with the suggested change.

brendan-m-murphy commented on tests/store/test_obssurface.py:
```diff
@@ -507,7 +509,7 @@ def test_read_noaa_obspack(bucket):
     assert data["ch4_variability"][0] == pytest.approx(2.093036e-09)
 
-
+@pytest.mark.xfail(reason="The values stored with _datasource_uuids are changing.")
```
This is no longer xfail, it's been changed to what Gareth did in PR #752

brendan-m-murphy commented on openghg/objectstore/metastore/_classic_metastore.py:
```diff
@@ -0,0 +1,159 @@
+
+
+def get_metakey(data_type: str) -> str:
+    """Return the metakey for a given data type."""
```
Done.

brendan-m-murphy commented on openghg/objectstore/metastore/_classic_metastore.py:
```diff
@@ -0,0 +1,159 @@
+    def from_bucket(cls: type[CM], bucket: str, data_type: str) -> CM:
+        """Initialise a ClassicMetaStore given a bucket and data type.
+
+        A ClassicMetaStore opened with this method must be closed to
```
Done.

brendan-m-murphy commented:
I've addressed most of the comments from the first round of reviews.

Unaddressed issues:
1. using `tmp_path`: I created an issue about converting tests to use the pytest built-ins like `tmp_path`
2. some xfail tests: I made a note in issue #727 that these need to be addressed
3. metakey info in `_classic_metastore.py` instead of central location: this has also been deferred to issue #727, which should make it so that this info isn't needed outside of a module in `objectstore`.

gareth-j reviewed (approved):
Looks good!
